File: tests/collectors/editors-ext.test.ts
import { test, expect, describe } from "bun:test";
import { parseExtensions, makeEditorsExtCollector } from "../../src/collectors/editors-ext";
import type { CollectorContext } from "../../src/collectors/types";
import { fakeEnv } from "../helpers/fake-env";

const ctx: CollectorContext = { redact: true, home: "/fake/home" };

describe("parseExtensions", () => {
  test("trims, drops blanks, sorts", () => {
    expect(parseExtensions("anthropic.claude-code\n\n  aaron-bond.better-comments \n")).toEqual([
      "aaron-bond.better-comments",
      "anthropic.claude-code",
    ]);
  });

  test("empty → []", () => {
    expect(parseExtensions("")).toEqual([]);
  });
});

describe("makeEditorsExtCollector", () => {
  test("collects VS Code and Cursor extensions into separate sections", async () => {
    const r = await makeEditorsExtCollector(
      fakeEnv({
        outputs: {
          code: "anthropic.claude-code\nadpyke.codesnap",
          cursor: "anthropic.claude-code",
        },
      }),
    )(ctx);
    expect(r["editor.vscode.extensions"]?.items).toEqual([
      { raw: "adpyke.codesnap", columns: ["adpyke.codesnap"] },
      { raw: "anthropic.claude-code", columns: ["anthropic.claude-code"] },
    ]);
    expect(r["editor.cursor.extensions"]?.items).toEqual([
      { raw: "anthropic.claude-code", columns: ["anthropic.claude-code"] },
    ]);
  });

  test("omits an editor that is not installed", async () => {
    const r = await makeEditorsExtCollector(fakeEnv({ outputs: { cursor: "some.ext" } }))(ctx);
    expect(r["editor.vscode.extensions"]).toBeUndefined();
    expect(r["editor.cursor.extensions"]).toBeDefined();
  });

  test("neither installed → {}", async () => {
    expect(await makeEditorsExtCollector(fakeEnv())(ctx)).toEqual({});
  });
});

File: internal/collect/editorsext_test.go
package collect

import (
	"reflect"
	"testing"
)

func TestParseExtensions(t *testing.T) {
	tests := []struct {
		name string
		in   string
		want []string
	}{
		{
			name: "trims, drops blanks, sorts",
			in:   "anthropic.claude-code\n\n  aaron-bond.better-comments \n",
			want: []string{"aaron-bond.better-comments", "anthropic.claude-code"},
		},
		{
			name: "empty -> empty",
			in:   "",
			want: []string{},
		},
		{
			name: "whitespace only -> empty",
			in:   "   \n\t\n  ",
			want: []string{},
		},
		{
			name: "single extension",
			in:   "some.ext",
			want: []string{"some.ext"},
		},
		{
			name: "multiple sorted",
			in:   "anthropic.claude-code\nadpyke.codesnap",
			want: []string{"adpyke.codesnap", "anthropic.claude-code"},
		},
		{
			name: "trailing newline only",
			in:   "anthropic.claude-code\n",
			want: []string{"anthropic.claude-code"},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := ParseExtensions(tt.in)
			if !reflect.DeepEqual(got, tt.want) {
				t.Errorf("ParseExtensions(%q) = %v, want %v", tt.in, got, tt.want)
			}
		})
	}
}
